app/src/main/java/nerd/tuxmobil/fahrplan/congress/settings/SettingsFragment.kt

```diff
@@ -80,6 +100,41 @@ class SettingsFragment : PreferenceFragment() {
         }
     }
 
+    override fun onActivityResult(requestCode: Int, resultCode: Int, data: Intent?) {
+        if (REQUEST_CODE_ALARM_TONE == requestCode && data != null) {
+            val alarmToneUri = data.getParcelableExtra<Uri>(RingtoneManager.EXTRA_RINGTONE_PICKED_URI)
+            onAlarmTonePicked(alarmToneUri)
+        } else {
+            super.onActivityResult(requestCode, resultCode, data)
+        }
+    }
+
+    private fun onAlarmTonePicked(alarmToneUri: Uri?) {
+        appRepository.updateAlarmToneUri(alarmToneUri)
+    }
+
+    override fun onPreferenceTreeClick(preference: Preference?): Boolean {
+        return if (getString(R.string.preference_key_alarm_tone) == preference?.key) {
+            onAlarmTonePreferenceClick()
+        } else {
+            super.onPreferenceTreeClick(preference)
+        }
+    }
+
+    private fun onAlarmTonePreferenceClick(): Boolean {
+        val alarmToneUri = appRepository.readAlarmToneUri()
+        val extras = bundleOf(
+                RingtoneManager.EXTRA_RINGTONE_TYPE to AlarmTonePreference.RINGTONE_TYPE,
+                RingtoneManager.EXTRA_RINGTONE_SHOW_DEFAULT to true,
+                RingtoneManager.EXTRA_RINGTONE_SHOW_SILENT to true,
+                RingtoneManager.EXTRA_RINGTONE_DEFAULT_URI to AlarmTonePreference.DEFAULT_VALUE_URI,
+                RingtoneManager.EXTRA_RINGTONE_EXISTING_URI to alarmToneUri
+        )
+        val intent = Intent(RingtoneManager.ACTION_RINGTONE_PICKER).apply { putExtras(extras) }
```

Comment:
Curious: Do you find your version easier to read than this?

```kotlin
val intent = Intent(RingtoneManager.ACTION_RINGTONE_PICKER).apply {
    putExtra(RingtoneManager.EXTRA_RINGTONE_TYPE, AlarmTonePreference.RINGTONE_TYPE)
    putExtra(RingtoneManager.EXTRA_RINGTONE_SHOW_DEFAULT, true)
    putExtra(RingtoneManager.EXTRA_RINGTONE_SHOW_SILENT, true)
    putExtra(RingtoneManager.EXTRA_RINGTONE_DEFAULT_URI, AlarmTonePreference.DEFAULT_VALUE_URI)
    putExtra(RingtoneManager.EXTRA_RINGTONE_EXISTING_URI, alarmToneUri)
}
```

If you prefer the `to` syntax I'd recommend creating an extension function like this:

```kotlin
fun Intent.withExtras(vararg pairs: Pair<String, Any?>) = putExtras(bundleOf(*pairs))
```

Then your calling code becomes

```kotlin
val intent = Intent(RingtoneManager.ACTION_RINGTONE_PICKER).withExtras(
    RingtoneManager.EXTRA_RINGTONE_TYPE to AlarmTonePreference.RINGTONE_TYPE,
    RingtoneManager.EXTRA_RINGTONE_SHOW_DEFAULT to true,
    RingtoneManager.EXTRA_RINGTONE_SHOW_SILENT to true,
    RingtoneManager.EXTRA_RINGTONE_DEFAULT_URI to AlarmTonePreference.DEFAULT_VALUE_URI,
    RingtoneManager.EXTRA_RINGTONE_EXISTING_URI to alarmToneUri
)
```

Reply:
:+1: Nice suggestion. I prepared a separate pull request in #313. Feel free to review and merge it.